lib/src/wiggle_abi/req_impl.rs

```diff
@@ -599,26 +604,38 @@ impl FastlyHttpReq for Session {
         if pending_req_handles.len() == 0 {
             return Err(Error::InvalidArgument);
         }
-        let targets = self.prepare_select_targets(&pending_req_handles.as_slice()?)?;
 
         // perform the select operation
-        let (fut_result, done_index, rest) = futures::future::select_all(targets).await;
-
-        // reinsert the other receivers before doing anything else, so they don't get dropped
-        self.reinsert_select_targets(rest);
-
-        let outcome = match fut_result {
-            Ok(resp) => {
-                let (resp_handle, body_handle) = self.insert_response(resp);
-                (done_index as u32, resp_handle, body_handle)
-            }
-            // Unfortunately, the ABI provides no means of returning error information
-            // from completed `select`.
-            Err(_) => (
-                done_index as u32,
-                INVALID_RESPONSE_HANDLE.into(),
-                INVALID_BODY_HANDLE.into(),
-            ),
+        let done_index = self
+            .select_impl(
+                pending_req_handles
+                    .as_slice()?
+                    .iter()
+                    .map(|handle| handle.to_owned().into()),
+            )
+            .await? as u32;
+
+        let item =
+            self.take_async_item(pending_req_handles.get(done_index).unwrap().read()?.into())?;
```

Comment:
```suggestion
            self.take_async_item(pending_req_handles[done_index].read()?.into())?;
```

Reply:
I thought this would work but turns out `pending_req_handles` is a &GuestPtr<[PendingReqHandle]> and we can't index directly. I tried out some other things but they were either longer or even more complex for little gain and so I think we need to just keep this unfortunately @aturon

Reply:
Ah ok, no worries!